### ISSUE TITLE
Allow WeakReference for interned objects

### ISSUE THREAD

**Issue by dair-targ** [immutables/immutables#759]
Add a flag to use weak references for `@Value.Immutable(intern = true)`. This could be done by providing an alternative way to initialize and use `INTERNER` at `Immutables.generate` and adding a way to get the values.

```
  private static final java.util.concurrent.ConcurrentHashMap<InternProxy, java.lang.ref.WeakReference<[type.typeImmutable.simple]>> INTERNER =
      new java.util.concurrent.ConcurrentHashMap<InternProxy, java.lang.ref.WeakReference<[type.typeImmutable.simple]>>();
  [else]
  private static final [guava].collect.Interner<InternProxy> INTERNER = [guava].collect.Interners.newWeakInterner();
  [/if]
```

This case is derived from machine-learning training application. Sometimes the training consists of steps during which the large number of intermediate "key" objects is being generated. Interning parts of those key objects helps to significantly reduce the memory footprint, however those "interned" parts would stay in the memory forever even if there's no way to link them from the outer world.

This could be a second-choice behavior as it represents another way to trade CPU time (required to handle WeakReferences) for lesser memory consumption.

**Comment by elucash** [comment]
Thank you for the feature request! I think we can add style option to switch to weak references. Previously, we decided to not implement this because workaround existed, but of course,
this would be more of custom code:

```java
@Value.Immutable abstract class WeaklyInternedObj {
    private static final Interner<WeaklyInternedObj> INTERNER = Interners.newWeakInterner();
    @Value.Check WeaklyInternedObj intern() {
       return INTERNER.intern(this);
    }
}
```

Unfortunately, I cannot say that this can be abstracted into a base class, the following example will not work due to a bug (just fixed 1e05f66ec0e607851a2fb5c3a23ad7cd4ec89a16 , not released yet)

```java
// This example is not working currently
abstract class BaseInterned<T extends BaseInterned<T>> {
  private static final Interner<Object> INTERNER = Interners.newStrongInterner();
  @SuppressWarnings("unchecked")
  @Value.Check
  protected T intern() {    // <-- this will not work, due to bug
    return (T) INTERNER.intern(this);
  }
}
@Value.Immutable
public abstract class CustomInternedNormalized extends BaseInterned<CustomInternedNormalized> {
  abstract int a();
  abstract int b();
}
```

However, `Style.weakInterning` will be introduced for this.

**Comment by dair-targ** [comment]
Thanks a lot for the prompt response and resolution!
